src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/MacProxy.cs

```diff
@@ -0,0 +1,125 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Net.Http;
+using System.Net;
+using System.Collections.Generic;
+using Microsoft.Win32.SafeHandles;
+using static Interop.CoreFoundation;
+using static Interop.RunLoop;
+
+using CFRunLoopRef = System.IntPtr;
+
+namespace System.Net.Http
+{
+    internal sealed class MacProxy : IWebProxy
+    {
+        public ICredentials Credentials
+        {
+            get => null;
+            set => throw new NotSupportedException();
+        }
+
+        static Uri GetProxyUri(string scheme, CFProxy proxy)
+        {
+            var uriBuilder = new UriBuilder(
+                scheme,
+                proxy.HostName,
+                proxy.PortNumber);
+
+            // FIXME: Handle user name and password
+
+            return uriBuilder.Uri;
+        }
+
+        public Uri ExecuteProxyAutoConfiguration(SafeCreateHandle cfurl, CFProxy proxy)
+        {
+            Uri result = null;
+            CFRunLoopRef runLoop = CFRunLoopGetCurrent();
+
+            // Callback that will be called after executing the configuration script
+            CFProxyAutoConfigurationResultCallback cb = (IntPtr client, IntPtr proxyListHandle, IntPtr error) =>
+            {
+                if (proxyListHandle != IntPtr.Zero)
+                {
+                    using (var proxyList = new SafeCFArrayHandle(proxyListHandle, false))
+                    {
+                        long proxyCount = CFArrayGetCount(proxyList);
+                        for (int i = 0; i < proxyCount; i++)
+                        {
+                            IntPtr proxyValue = CFArrayGetValueAtIndex(proxyList, i);
+                            using (SafeCFDictionaryHandle proxyDict = new SafeCFDictionaryHandle(proxyValue, false))
+                            {
+                                CFProxy proxy = new CFProxy(proxyDict);
+                                if (proxy.ProxyType == CFProxy.kCFProxyTypeHTTP || proxy.ProxyType == CFProxy.kCFProxyTypeHTTPS)
+                                {
+                                    result = GetProxyUri("http", proxy);
+                                    break;
+                                }
+                            }
+                        }
+                    }
+                }
+                CFRunLoopStop(runLoop);
+            };
+
+            var clientContext = new CFStreamClientContext();
```

Comment:
Is it allowed for Apple to change the size/shape of CFStreamClientContext?  I looked at the docs, and while they list properties, they don't detail the actual shape of this.  If it were to change in size, we'd potentially end up corrupting the stack.

Reply:
It's a struct, so no, they are not allowed to change the shape. This is compatible with the C definition.

Reply:
Excellent.